library/ssl_msg.c

```diff
@@ -2033,31 +2033,6 @@ int mbedtls_ssl_fetch_input( mbedtls_ssl_context *ssl, size_t nb_want )
          * header) and/or some other records in the same datagram.
          */
 
-        /*
```

Comment:
It took me a while to convince myself that the commit “Move logic for consumption of DTLS record within datagram” was ok. I know that the complexity of the code is what you're working on solving with MPS, but even so, I'd appreciate more documentation about the existing code. The commit message was ok to understand the purpose and method of the commit, but insufficient to understand why the commit is correct.

This commit moves a chunk of code from near the beginning of `mbedtls_ssl_fetch_input` to the end of `ssl_consume_current_message`. `ssl_consume_current_message` is called in a single place, from `mbedtls_ssl_read_record`. `mbedtls_ssl_fetch_input` is called in several places.

In the code path in `mbedtls_ssl_read_record`, this means that the chunk of code is moved up past three things:

* Checks at the beginning of `mbedtls_ssl_fetch_input`. This is ok because these checks are orthogonal to what the moved code does.
* The call to `ssl_load_buffered_record` at the beginning of `ssl_get_next_record`. This is ok because `ssl_load_buffered_record` does nothing if there is at least one record left (via `ssl_next_record_is_in_datagram`), whereas the moved code only does something if there is at least one more record left.
* The call to `ssl_load_buffered_message` in `mbedtls_ssl_read_record`. This is ok because `ssl_load_buffered_message` is only called if there is at least one record left (via `ssl_next_record_is_in_datagram`), whereas the moved code only does something if there is at least one more record left.

In `ssl_get_next_record` and `ssl_parse_client_hello_v2`, `mbedtls_ssl_fetch_input` is only called for non-DTLS. Since the moved code only concerns DTLS, this is ok.

In `ssl_parse_client_hello`, `mbedtls_ssl_fetch_input` is called in two places. In both cases, it's only called for the initial ClientHello, not when renegotiating. The initial ClientHello is the first message in the connection, so it must be the very first record, and therefore there cannot have been a previous record that needs to be moved from, and the moved code would never trigger.